rules-reviewed/camel3/camel2/xml-removed-components.windup.groovy

```diff
@@ -0,0 +1,82 @@
+import org.jboss.windup.rules.files.condition.FileContent
+import org.ocpsoft.rewrite.config.Condition
+import org.ocpsoft.rewrite.context.EvaluationContext
+
+import java.util.function.BiFunction
+import java.util.function.Function
+
+final String FROM_XML_FILES_IN_PROJECT = "xmlFilesInProject"
+final String FROM_FILES_IN_PROJECT = "filesInProject"
+final Set<String> componentsAdded = ["aws-cw", "aws-ddb", "aws-ec2", "aws-iam", "aws-kinesis", "aws-kms", "aws-lambda", "aws-mq", "aws-s3", "aws-sdb", "aws-ses", "aws-sns", "aws-sqs", "aws-swf", "aws-xray"]
+
+final IssueCategory mandatoryIssueCategory = new IssueCategoryRegistry().getByID(IssueCategoryRegistry.MANDATORY)
+final Link awsLink = Link.to("Camel 3 - Migration Guide: AWS", "https://camel.apache.org/manual/latest/camel-3-migration-guide.html#_aws")
+
+final BiFunction<String, String, Boolean> isThereArtifactId = { String xmlDependenciesBlock, String component -> 
+    xmlDependenciesBlock.contains("<artifactId>camel-$component</artifactId>" as CharSequence) }
+
+final BiFunction<String, String, Condition> xmlCondition = { String component, String namespace -> 
+    XmlFile.from(FROM_XML_FILES_IN_PROJECT)
+            .matchesXpath("//*/c:route/*[starts-with(@uri, '$component:')]")
+            .namespace("c", namespace)
+            .as("$component-$namespace")
+}
+
+final Function<String, Condition> javaCondition = { String component ->
+    FileContent.from(FROM_FILES_IN_PROJECT)
+        .matches("(\"$component:")
+        .inFileNamed("{*}.java")
+        .as("$component-java")
+}
+
+ruleSet("xml-removed-components-groovy")
+    .addSourceTechnology(new TechnologyReference("camel", "[2,3)"))
+    .addTargetTechnology(new TechnologyReference("camel", "[3,)"))
+    .addRule()
+    .when(
+        XmlFile.matchesXpath("/m:project/m:dependencies[m:dependency/m:groupId/text() = 'org.apache.camel' and m:dependency/m:artifactId/text() = 'camel-aws']")
+            .inFile("pom.xml").namespace("m", "http://maven.apache.org/POM/4.0.0")
+    )
+    .perform(new AbstractIterationOperation<XmlTypeReferenceModel>()
+    {
+        void perform(GraphRewrite event, EvaluationContext context, XmlTypeReferenceModel payload) {
+            final FileModel fileModel = payload.getFile()
+            final String filePath = StringUtils.removeEnd(fileModel.getFilePath(), fileModel.getFileName())
+            Query.fromType(XmlFileModel.class).withProperty(FileModel.FILE_PATH, QueryPropertyComparisonType.CONTAINS_TOKEN, filePath).as(FROM_XML_FILES_IN_PROJECT).evaluate(event, context)
+            Query.fromType(FileModel.class).withProperty(FileModel.FILE_PATH, QueryPropertyComparisonType.CONTAINS_TOKEN, filePath).as(FROM_FILES_IN_PROJECT).evaluate(event, context)
+            final String xmlDependenciesBlock = payload.getSourceSnippit()
+            componentsAdded.stream()
+                .filter { component -> !isThereArtifactId.apply(xmlDependenciesBlock, component) }
```

Comment:
So, the query is : 
for all the xml files with dependency `camel-aws`
   for all the `aws-*` components
      check if that component NEW artifact `camel-aws-*` is NOT in the fat-jar     
      
When this would happen if you are analising a Camel 2 app ?

Reply:
The Camel rules are intended to support the adoption of the latest Camel 3 release.
This could mean having to run a Camel 2.X to 3.0 migration and then a Camel 3 minor version upgrade: in this second case, checking that the dependency has not been added yet, would avoid having false positive from this rule afaik.

Reply:
Sure, but how can you have `org.apache.camel:camel-aws` (so the Camel 2 one, including the S3 adapter) as a dependency and also having in the classpath the new `org.apache.camel:camel-aws-s3` ?

Reply:
If you could provide test data, it would help a lot to fix the case you're reporting.

Reply:
+1 for a test case. Note that it's `org.apache.camel.springboot:camel-aws-s3` in Camel 3

Reply:
I'm not sure I understand. I was questioning if it's possible to have camel 2 and camel 3 at the same time in the same project.   
Is it possible to have both ?

Reply:
Ok, let's try to summarize this thread as well.
@jonathanvila reported it's impossible (at least to my understanding from this discussion) to have both Camel 2 and 3 artifacts so, since `org.apache.camel:camel-aws` has been removed in Camel 3 (there's an edge case since it has been released with Camel `3.0.0-M1` but i don't think it's so edge to not take it into account) it represents a "good enough" condition without any need for further checks (i.e. `isThereArtifactId`).

It seems to me a good point so I want to fix this in this PR because I think it's due to the performance change in condition from the XML original rule.

If @mmelko and @johnpoth i don't hear nothing from you, I take this as a silent consent (but obviously feel free to stop me :wink:)

Reply:
Applied in commit https://github.com/windup/windup-rulesets/pull/436/commits/40223177d50b7e1d778a9c6259449340b80764da